documentation/configuration.md
# Configuration

Liftbridge provides limited configuration through command-line flags and full
configuration using a configuration file. Flags will always take precedent
over settings in the configuration file.

The configuration file is passed in using the `--config` flag:

```
$ liftbridge --config liftbridge.conf
```

To get a full list of the CLI flags, use the `--help` flag:

```
$ liftbridge --help
NAME:
   liftbridge - Lightweight, fault-tolerant message streams

USAGE:
   liftbridge [global options] command [command options] [arguments...]

VERSION:
   0.0.1

COMMANDS:
     help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --config FILE, -c FILE                      load configuration from FILE
   --server-id value, --id value               ID of the server in the cluster if there is no stored ID (default: random ID)
   --namespace value, --ns value               cluster namespace (default: "liftbridge-default")
   --nats-servers ADDR[,ADDR], -n ADDR[,ADDR]  connect to NATS cluster at ADDR[,ADDR] (default: "nats://localhost:4222")
   --data-dir DIR, -d DIR                      store data in DIR (default: "/tmp/liftbridge/<namespace>")
   --port value, -p value                      port to bind to (default: 9292)
   --tls-cert value                            server certificate file
   --tls-key value                             private key for server certificate
   --level value, -l value                     logging level [debug|info|warn|error] (default: "info")
   --raft-bootstrap-seed                       bootstrap the Raft cluster by electing self as leader if there is no existing state
   --raft-bootstrap-peers value                bootstrap the Raft cluster with the provided list of peer IDs if there is no existing state
   --help, -h                                  show help
   --version, -v                               print the version
```

## Configuration File Format

The configuration file format borrows from the [NATS configuration
format](https://nats.io/documentation/server/gnatsd-config/). It is a flexible
format that combines the best of traditional configuration formats and newer
styles such as JSON and YAML.

The config file format supports the following syntax:

- Mixed Arrays: `[...]`
- Nested Maps: `{...}`
- Multiple comment types: `#` and `//`
- Key value assigments using:
  - Equals sign (`foo = 2`)
  - Colon (`foo: 2`)
  - Whitespace (`foo 2`)
- Maps can be assigned with no key separator
- Semicolons as value terminators in key/value assignments are optional

## Example Configuration File

An example configuration file is shown below.

```
listen: localhost:9293
data.dir: /tmp/liftbridge/server-2
log.level: debug

# Define NATS cluster to connect to.
nats {
    servers: ["nats://localhost:4300", "nats://localhost:4301"]
}

# Specify message log settings.
log {
    retention.max.age: "24h"
}

# Specify cluster settings.
clustering {
    server.id: server-2
    raft.logging: true
    raft.bootstrap.seed: true
    replica.max.lag.time: "20s"
}
```

## Configuration Settings

Below is the list of Liftbridge configuration settings, including the name of
the setting in the configuration file and the CLI flag if it exists.

| Name | Flag | Description | Type | Default | Valid Values |
|:----|:----|:----|:----|:----|:----|
| listen | | The server listen host/port. Also used for the connection address if host is not set. | string | localhost:9292  | |
| host | | The server host. | string | localhost | |
| port | port | The server port. | int | 9292 | |
| tls.key | tls-key | The private key file for server certificate. This must be set in combination with `tls.cert` to enable TLS. | string | |
| tls.cert | tls-cert | The server certificate file. This must be set in combination with `tls.key` to enable TLS. | string | |
| log.level | level | The logging level. | string | info | [debug, info, warn, error] |
| log.recovery | | Log messages resulting from the replay of the Raft log on server recovery. | bool | false | |
| data.dir | data-dir | The directory to store data in. | string | /tmp/liftbridge/namespace | |
| batch.max.messages | | The maximum number of messages to batch when writing to disk. | int | 1024 |
| batch.wait.time | | The time to wait to batch more messages when writing to disk. | duration | 0 | |
| metadata.cache.max.age | | The maximum age of cached broker metadata. | duration | 2m | |
| nats | | NATS configuration. | map | | [See below](#nats-configuration-settings) |
| log | | Stream write-ahead log configuration. | map | | [See below](#log-configuration-settings) |
| clustering | | Broker cluster configuration. | map | | [See below](#cluster-configuration-settings) |

### NATS Configuration Settings

Below is the list of the configuration settings for the `nats` part of
the configuration file.

| Name | Flag | Description | Type | Default | Valid Values |
|:----|:----|:----|:----|:----|:----|
| servers | nats-servers | List of NATS hosts to connect to. | list | nats://localhost:4222 | |

### Log Configuration Settings

Below is the list of the configuration settings for the `log` part of
the configuration file.

| Name | Flag | Description | Type | Default | Valid Values |
|:----|:----|:----|:----|:----|:----|
| retention.max.bytes | | The maximum size a stream's log can grow to, in bytes, before we will discard old log segments to free up space. A value of 0 indicates no limit. | int64 | 0 | |
| retention.max.messages | | The maximum size a stream's log can grow to, in number of messages, before we will discard old log segments to free up space. A value of 0 indicates no limit. | int64 | 0 | |
| retention.max.age | | The TTL for stream log segment files, after which they are deleted. A value of 0 indicates no TTL. | duration | 168h | |
| cleaner.interval | | The frequency to check if a new stream log segment file should be rolled and whether any segments are eligible for deletion based on the retention policy or compaction if enabled. | duration | 5m | |
| log.roll.time | | The maximum time before a new stream log segment is rolled out. A value of 0 means new segments will only be rolled when `segment.max.bytes` is reached. Retention is always done a file at a time, so a larger value means fewer files but less granular control over retention. | duration | value of `retention.max.age` | |
| segment.max.bytes | | The maximum size of a single stream log segment file in bytes. Retention is always done a file at a time, so a larger segment size means fewer files but less granular control over retention. | int64 | 268435456 | |
| compact | | Enables stream log compaction. Compaction works by retaining only the latest message for each key and discarding older messages. The frequency in which compaction runs is controlled by `cleaner.interval`. | bool | false | |
| compact.max.goroutines | | The maximum number of concurrent goroutines to use for compaction on a stream log (only applicable if `compact` is enabled). | int | 10 | |

### Clustering Configuration Settings

Below is the list of the configuration settings for the `clustering` part of
the configuration file.

| Name | Flag | Description | Type | Default | Valid Values |
|:----|:----|:----|:----|:----|:----|
| server.id | server-id | ID of the server in the cluster. | string | random id | string with no spaces or periods |
| namespace | namespace | Cluster namespace. | string | liftbridge-default | string with no spaces or periods |
| raft.snapshot.retain | | The number Raft log snapshots to retain on disk. | int | 2 | |
| raft.snapshot.threshold | | Controls how many outstanding logs there must be before taking a snapshot. This prevents excessive snapshots when a small set of logs can be replayed. | int | 8192 | |
| raft.cache.size | | The number of Raft logs to hold in memory for quick lookup. | int | 512 | |
| raft.bootstrap.seed | raft-bootstrap-seed | Bootstrap the Raft cluster by electing self as leader if there is no existing state. | bool | false | |
| raft.bootstrap.peers | raft-bootstrap-peers | Bootstrap the Raft cluster with the provided list of peer IDs if there is no existing state. | list | | |
| raft.logging | | Enables logging in the Raft subsystem. | bool | false | |
| replica.max.lag.time | | If a follower hasn't sent any replication requests or hasn't caught up to the leader's log end offset for at least this time, the leader will remove the follower from ISR. | duration | 10s | |
| replica.max.leader.timeout | | If a leader hasn't sent any replication responses for at least this time, the follower will report the leader to the controller. If a majority of the replicas report the leader, a new leader is selected by the controller. | duration | 10s | |
| replica.fetch.timeout | | Timeout duration for follower replication requests. | duration | 3s | |
| min.insync.replicas | | Specifies the minimum number of replicas that must acknowledge a stream write before it can be committed. If the ISR drops below this size, messages cannot be committed. | int | 1 | [1,...] |


server/config.go
package server

import (
	"fmt"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/dustin/go-humanize"
	"github.com/hako/durafmt"
	"github.com/nats-io/nats.go"
	"github.com/nats-io/nuid"
	log "github.com/sirupsen/logrus"

	"github.com/liftbridge-io/liftbridge/server/conf"
)

const (
	// DefaultNamespace is the default cluster namespace to use if one is not
	// specified.
	DefaultNamespace = "liftbridge-default"

	// DefaultPort is the port to bind to if one is not specified.
	DefaultPort = 9292
)

const (
	defaultListenAddress           = "0.0.0.0"
	defaultConnectionAddress       = "localhost"
	defaultReplicaMaxLagTime       = 10 * time.Second
	defaultReplicaMaxLeaderTimeout = 10 * time.Second
	defaultRaftSnapshots           = 2
	defaultRaftCacheSize           = 512
	defaultMetadataCacheMaxAge     = 2 * time.Minute
	defaultBatchMaxMessages        = 1024
	defaultReplicaFetchTimeout     = 3 * time.Second
	defaultMinInsyncReplicas       = 1
	defaultRetentionMaxAge         = 7 * 24 * time.Hour
	defaultCleanerInterval         = 5 * time.Minute
	defaultMaxSegmentBytes         = 1024 * 1024 * 256 // 256MB
	defaultLogRollTime             = defaultRetentionMaxAge
	defaultCompactMaxGoroutines    = 10
)

// LogConfig contains settings for controlling the message log for a stream.
type LogConfig struct {
	RetentionMaxBytes    int64
	RetentionMaxMessages int64
	RetentionMaxAge      time.Duration
	CleanerInterval      time.Duration
	SegmentMaxBytes      int64
	LogRollTime          time.Duration
	Compact              bool
	CompactMaxGoroutines int
}

// RetentionString returns a human-readable string representation of the
// retention policy.
func (l LogConfig) RetentionString() string {
	str := "["
	prefix := ""
	if l.RetentionMaxMessages != 0 {
		str += fmt.Sprintf("Messages: %s", humanize.Comma(l.RetentionMaxMessages))
		prefix = ", "
	}
	if l.RetentionMaxBytes != 0 {
		str += fmt.Sprintf("%sSize: %s", prefix, humanize.IBytes(uint64(l.RetentionMaxBytes)))
		prefix = ", "
	}
	if l.RetentionMaxAge > 0 {
		str += fmt.Sprintf("%sAge: %s", prefix, durafmt.Parse(l.RetentionMaxAge))
		prefix = ", "
	}
	if prefix == "" {
		str += "no limits"
	}
	str += fmt.Sprintf(", Compact: %t", l.Compact)
	str += "]"
	return str
}

// ClusteringConfig contains settings for controlling cluster behavior.
type ClusteringConfig struct {
	ServerID                string
	Namespace               string
	RaftSnapshots           int
	RaftSnapshotThreshold   uint64
	RaftCacheSize           int
	RaftBootstrapSeed       bool
	RaftBootstrapPeers      []string
	RaftLogging             bool
	ReplicaMaxLagTime       time.Duration
	ReplicaMaxLeaderTimeout time.Duration
	ReplicaFetchTimeout     time.Duration
	MinISR                  int
}

// Config contains all settings for a Liftbridge Server.
type Config struct {
	Listen              HostPort
	Host                string
	Port                int
	LogLevel            uint32
	LogRecovery         bool
	LogSilent           bool
	DataDir             string
	BatchMaxMessages    int
	BatchWaitTime       time.Duration
	MetadataCacheMaxAge time.Duration
	TLSKey              string
	TLSCert             string
	NATS                nats.Options
	Log                 LogConfig
	Clustering          ClusteringConfig
}

// NewDefaultConfig creates a new Config with default settings.
func NewDefaultConfig() *Config {
	config := &Config{
		NATS: nats.GetDefaultOptions(),
		Port: DefaultPort,
	}
	config.LogLevel = uint32(log.InfoLevel)
	config.BatchMaxMessages = defaultBatchMaxMessages
	config.MetadataCacheMaxAge = defaultMetadataCacheMaxAge
	config.Clustering.ServerID = nuid.Next()
	config.Clustering.Namespace = DefaultNamespace
	config.Clustering.ReplicaMaxLagTime = defaultReplicaMaxLagTime
	config.Clustering.ReplicaMaxLeaderTimeout = defaultReplicaMaxLeaderTimeout
	config.Clustering.ReplicaFetchTimeout = defaultReplicaFetchTimeout
	config.Clustering.RaftSnapshots = defaultRaftSnapshots
	config.Clustering.RaftCacheSize = defaultRaftCacheSize
	config.Clustering.MinISR = defaultMinInsyncReplicas
	config.Log.SegmentMaxBytes = defaultMaxSegmentBytes
	config.Log.RetentionMaxAge = defaultRetentionMaxAge
	config.Log.LogRollTime = defaultLogRollTime
	config.Log.CleanerInterval = defaultCleanerInterval
	return config
}

// GetListenAddress returns the address and port to listen to.
func (c Config) GetListenAddress() HostPort {
	if len(c.Listen.Host) > 0 {
		return c.Listen
	}

	if len(c.Host) > 0 {
		return HostPort{
			Host: c.Host,
			Port: c.Port,
		}
	}

	return HostPort{
		Host: defaultListenAddress,
		Port: DefaultPort,
	}
}

// GetConnectionAddress returns the host if specified and listen otherwise.
func (c Config) GetConnectionAddress() HostPort {
	if len(c.Host) > 0 {
		return HostPort{
			Host: c.Host,
			Port: c.Port,
		}
	}

	return HostPort{
		Host: defaultConnectionAddress,
		Port: DefaultPort,
	}
}

// GetLogLevel converts the level string to its corresponding int value. It
// returns an error if the level is invalid.
func GetLogLevel(level string) (uint32, error) {
	var l uint32
	switch strings.ToLower(level) {
	case "debug":
		l = uint32(log.DebugLevel)
	case "info":
		l = uint32(log.InfoLevel)
	case "warn":
		l = uint32(log.WarnLevel)
	case "error":
		l = uint32(log.ErrorLevel)
	default:
		return 0, fmt.Errorf("Invalid log.level setting %q", level)
	}
	return l, nil
}

// NewConfig creates a new Config with default settings and applies any
// settings from the given configuration file.
func NewConfig(configFile string) (*Config, error) {
	config := NewDefaultConfig()

	if configFile == "" {
		return config, nil
	}
	c, err := conf.ParseFile(configFile)
	if err != nil {
		return nil, err
	}

	// Reset LogRollTime since this will get overwritten later.
	config.Log.LogRollTime = 0

	for k, v := range c {
		switch strings.ToLower(k) {
		case "listen":
			hp, err := parseListen(v)
			if err != nil {
				return nil, err
			}

			config.Listen = *hp
		case "port":
			config.Port = int(v.(int64))
		case "host":
			config.Host = v.(string)
		case "log.level":
			level, err := GetLogLevel(v.(string))
			if err != nil {
				return nil, err
			}
			config.LogLevel = level
		case "log.recovery":
			config.LogRecovery = v.(bool)
		case "data.dir":
			config.DataDir = v.(string)
		case "batch.max.messages":
			config.BatchMaxMessages = int(v.(int64))
		case "batch.wait.time":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return nil, err
			}
			config.BatchWaitTime = dur
		case "metadata.cache.max.age":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return nil, err
			}
			config.MetadataCacheMaxAge = dur
		case "tls.key":
			config.TLSKey = v.(string)
		case "tls.cert":
			config.TLSCert = v.(string)
		case "nats":
			if err := parseNATSConfig(v.(map[string]interface{}), &config.NATS); err != nil {
				return nil, err
			}
		case "log":
			if err := parseLogConfig(config, v.(map[string]interface{})); err != nil {
				return nil, err
			}
		case "clustering":
			if err := parseClusteringConfig(config, v.(map[string]interface{})); err != nil {
				return nil, err
			}
		default:
			return nil, fmt.Errorf("Unknown configuration setting %q", k)
		}
	}

	// If LogRollTime is not set, default it to the retention time.
	if config.Log.LogRollTime == 0 {
		config.Log.LogRollTime = config.Log.RetentionMaxAge
	}

	return config, nil
}

// parseNATSConfig parses the `nats` section of a config file and populates the
// given nats.Options.
func parseNATSConfig(m map[string]interface{}, opts *nats.Options) error {
	for k, v := range m {
		switch strings.ToLower(k) {
		case "servers":
			servers := v.([]interface{})
			opts.Servers = make([]string, len(servers))
			for i, p := range servers {
				opts.Servers[i] = p.(string)
			}
		default:
			return fmt.Errorf("Unknown nats configuration setting %q", k)
		}
	}
	return nil
}

// parseLogConfig parses the `log` section of a config file and populates the
// given Config.
func parseLogConfig(config *Config, m map[string]interface{}) error {
	for k, v := range m {
		switch strings.ToLower(k) {
		case "retention.max.bytes":
			config.Log.RetentionMaxBytes = v.(int64)
		case "retention.max.messages":
			config.Log.RetentionMaxMessages = v.(int64)
		case "retention.max.age":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Log.RetentionMaxAge = dur
		case "cleaner.interval":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Log.CleanerInterval = dur
		case "segment.max.bytes":
			config.Log.SegmentMaxBytes = v.(int64)
		case "log.roll.time":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Log.LogRollTime = dur
		case "compact":
			config.Log.Compact = v.(bool)
		case "compact.max.goroutines":
			config.Log.CompactMaxGoroutines = v.(int)
		default:
			return fmt.Errorf("Unknown log configuration setting %q", k)
		}
	}
	return nil
}

// parseClusteringConfig parses the `clustering` section of a config file and
// populates the given Config.
func parseClusteringConfig(config *Config, m map[string]interface{}) error {
	for k, v := range m {
		switch strings.ToLower(k) {
		case "server.id":
			config.Clustering.ServerID = v.(string)
		case "namespace":
			config.Clustering.Namespace = v.(string)
		case "raft.snapshot.retain":
			config.Clustering.RaftSnapshots = int(v.(int64))
		case "raft.snapshot.threshold":
			config.Clustering.RaftSnapshotThreshold = uint64(v.(int64))
		case "raft.cache.size":
			config.Clustering.RaftCacheSize = int(v.(int64))
		case "raft.bootstrap.seed":
			config.Clustering.RaftBootstrapSeed = v.(bool)
		case "raft.bootstrap.peers":
			peers := v.([]interface{})
			config.Clustering.RaftBootstrapPeers = make([]string, len(peers))
			for i, p := range peers {
				config.Clustering.RaftBootstrapPeers[i] = p.(string)
			}
		case "raft.logging":
			config.Clustering.RaftLogging = v.(bool)
		case "replica.max.lag.time":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Clustering.ReplicaMaxLagTime = dur
		case "replica.max.leader.timeout":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Clustering.ReplicaMaxLeaderTimeout = dur
		case "replica.fetch.timeout":
			dur, err := time.ParseDuration(v.(string))
			if err != nil {
				return err
			}
			config.Clustering.ReplicaFetchTimeout = dur
		case "min.insync.replicas":
			config.Clustering.MinISR = int(v.(int64))
		default:
			return fmt.Errorf("Unknown clustering configuration setting %q", k)
		}
	}
	return nil
}

// HostPort is simple struct to hold parsed listen/addr strings.
type HostPort struct {
	Host string
	Port int
}

// parseListen will parse the `listen` option containing the host and port.
func parseListen(v interface{}) (*HostPort, error) {
	hp := &HostPort{}
	switch v.(type) {
	// Only a port
	case int64:
		hp.Port = int(v.(int64))
	case string:
		host, port, err := net.SplitHostPort(v.(string))
		if err != nil {
			return nil, fmt.Errorf("Could not parse address string %q", v)
		}
		hp.Port, err = strconv.Atoi(port)
		if err != nil {
			return nil, fmt.Errorf("Could not parse port %q", port)
		}
		hp.Host = host
	}
	return hp, nil
}
